src/seals/base_envs.py

```diff
@@ -247,6 +346,88 @@ def feature_matrix(self):
         # Construct lazily to save memory in algorithms that don't need features.
         if self._feature_matrix is None:
+            # TODO(juan) Space() does not have an `n` attribute (?).
+            #  Are we hinting the wrong type?
             n_states = self.state_space.n
             self._feature_matrix = np.eye(n_states)
         return self._feature_matrix
+
+    @property
+    def state_dim(self):
+        """Number of states in this MDP (int)."""
+        return self.transition_matrix.shape[0]
+
+    @property
+    def action_dim(self) -> int:
+        """Number of action vectors (int)."""
+        return self.transition_matrix.shape[1]
+
+    @property
+    def obs_dim(self) -> int:
+        """Size of observation vectors for this MDP."""
+        return self.observation_matrix.shape[1]
+
+    @property
+    def obs_dtype(self) -> int:
+        """Data type of observation vectors (e.g. np.float32)."""
+        return self.observation_matrix.dtype
+
+
+class TabularModelPOMDP(BaseTabularModelPOMDP[np.ndarray]):
+    """Tabular model POMDP.
+
+    This class is specifically for environments where observation != state,
+    from both a typing perspective but also by defining the method that
+    draws observations from the state.
+
+    The tabular model is deterministic in drawing observations from the state,
+    in that given a certain state, the observation is always the same;
+    a vector with self.obs_dim entries.
+    """
+
+    def obs_from_state(self, state: int) -> np.ndarray:
+        """Computes observation from state."""
+        # Copy so it can't be mutated in-place (updates will be reflected in
+        # self.observation_matrix!)
+        obs = self.observation_matrix[state].copy()
+        assert obs.ndim == 1, obs.shape
+        return obs
+
+
+class TabularModelMDP(BaseTabularModelPOMDP[int]):
+    """Tabular model MDP.
+
+    A tabular model MDP is a tabular MDP where the transition and reward
+    matrices are constant.
+    """
+
+    def __init__(
+        self,
+        *,
+        transition_matrix: np.ndarray,
+        reward_matrix: np.ndarray,
+        horizon: float = np.inf,
+        initial_state_dist: Optional[np.ndarray] = None,
+    ):
+        """Initializes a tabular model MDP.
+
+        Args:
+            transition_matrix: Matrix of shape `(n_states, n_actions, n_states)`
+                containing transition probabilities.
+            reward_matrix: Matrix of shape `(n_states, n_actions, n_states)`
+                containing reward values.
+            initial_state_dist: Distribution over initial states. Shape `(n_states,)`.
+            horizon: Maximum number of steps to take in an episode.
+        """
+        super().__init__(
+            transition_matrix=transition_matrix,
+            reward_matrix=reward_matrix,
+            horizon=horizon,
+            initial_state_dist=initial_state_dist,
+            observation_matrix=np.eye(transition_matrix.shape[0]),
```

Comment:
Something seems off here. We're basically giving a dummy observation matrix to the code, that never gets used because of the `obs_from_state` override, and doesn't even produce the same values (one-hot coded vs integer index).

If we make change I suggested above to remove `observation_matrix` from `BaseTabularModelPOMDP`, you could just switch this to inherit directly from `BaseTabularModelPOMDP` and get rid of the `observation_matrix` here entirely. You'd probably need to make `BaseTabularModelPOMDP` take `observation_space` as an argument (specifying `obs_dim` and `obs_dtype` won't cut it if you want it to be discrete...), but that seems like a reasonable choice, then just move the current `construct_obs_space` logic into `TabularModelPOMDP`.

I'm not 100% satisfied with that, as it does seem like we'd probably want `TabularModelMDP` to be-a `TabularModelPOMDP`, but it doesn't seem like a major problem if they're both concrete classes and so specialized in different ways.

Alternatively if you wanted to keep the current hierarchy, you could just make `observation_matrix` not bogus. Either keep it as-is and delete `obs_from_state` (you get one hot vectors, which is OK) or change `np.eye` to `np.arange` (the observation space would be a bit weird there though).